rails/spec/controllers/admin/project_links_controller_spec.rb

```diff
@@ -0,0 +1,325 @@
+        expect(get :new).to render_template('new')
+        # Redirect, and show error when not allowed:
+      end
+      it 'should display all the projects in the project dropdown' do
+        get :new
+        expect(assigns(:projects)).to include(project_1)
+        expect(assigns(:projects)).to include(project_2)
+        expect(assigns(:projects)).to include(project_3)
+      end
+      it 'should return an OK http status' do
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+    describe 'Edit' do
+      before(:each) do
+        allow(Admin::ProjectLink).to receive(:find).and_return(@link_1)
+      end
+      it 'it should show the Edit form' do
+        expect(get :edit).to render_template('edit')
+        # Redirect, and show error when not allowed:
+      end
+      it 'should display all the projects in the project dropdown' do
+        get :edit
+        expect(assigns(:projects)).to include(project_1)
+        expect(assigns(:projects)).to include(project_2)
+        expect(assigns(:projects)).to include(project_3)
+      end
+      it 'should return an OK http status' do
+        expect(response).to have_http_status(:ok)
+      end
+    end
+
+      describe 'for a link in project 2' do
+        let(:project_id) { project_2.id }
+        it 'should let them' do
+          post :create, params
+          link = assigns(:project_link)
+          expect(assigns(:project_link)).to be_valid
+          expect(response).to redirect_to(admin_project_link_path(link))
+        end
+      end
+    end
+  end
+
+  describe 'Nested controller methods' do
+    let(:user) { admin_user }
+    let(:project) { project_2 }
+
+    context 'when nested under project route' do
+      context 'get index' do
+        it 'should restrict the list of links to that project' do
+          get 'index', project_id: project.id
+          expect(assigns('project_links')).to include(@link_2)
+          expect(assigns('project_links')).not_to include(@link_1)
+          expect(assigns('project_links')).not_to include(@link_3)
+          expect(response).to have_http_status(:ok)
+        end
+      end
+      context 'get new' do
+        it 'should only display the selected project in the drop down' do
+          get 'new', project_id: project.id
+          expect(assigns('projects')).to include(project_2)
+          expect(assigns('projects')).not_to include(project_1)
+          expect(assigns('projects')).not_to include(project_3)
+          expect(response).to have_http_status(:ok)
+        end
+
+        it 'should pre-select the correct project from the route' do
+          [project_1, project_2, project_3].each do |chosen_project|
+            get 'new', project_id: chosen_project.id
+            new_project_link = assigns('project_link')
+            expect(new_project_link.project_id).to eq(chosen_project.id)
+            expect(response).to have_http_status(:ok)
+          end
+        end
+      end
```

Comment:
It'd be helpful to have a test here to make sure project admins can still create and update projects even when doing so in a nested path.  For example with the current code if we changed the project authorization rules it could break this code path.